File: hw1/static/Sudoku.cpp
#include <iostream>
#include <vector>
#include <set>

class Grid {
public:
    const int BOX_SIZE = 3;
    const int GRID_SIZE = 9;
    std::vector<std::vector<int>> grid;

    Grid() {
        grid.resize(GRID_SIZE, std::vector<int>(GRID_SIZE, 0));
    }

    std::vector<int> getRow(int row) {
        return grid[row];
    }

    std::vector<int> getColumn(int col) {
        std::vector<int> column;
        for (int i = 0; i < GRID_SIZE; ++i) {
            column.push_back(grid[i][col]);
        }
        return column;
    }

    std::vector<int> getBox(int row, int col) {
        std::vector<int> box;
        int box_row_start = (row / BOX_SIZE) * BOX_SIZE;
        int box_col_start = (col / BOX_SIZE) * BOX_SIZE;
        for (int i = box_row_start; i < box_row_start + BOX_SIZE; ++i) {
            for (int j = box_col_start; j < box_col_start + BOX_SIZE; ++j) {
                box.push_back(grid[i][j]);
            }
        }
        return box;
    }
};

class Sudoku : public Grid {
public:
    Sudoku(std::string input_str) {
        parse(input_str);
    }

    void parse(std::string input_str) {
        for (int row = 0; row < GRID_SIZE; ++row) {
            for (int col = 0; col < GRID_SIZE; ++col) {
                int index = row * GRID_SIZE + col;
                int value = input_str[index] - '0';
                grid[row][col] = value;
            }
        }
    }

    std::vector<std::vector<std::set<int>>> getInference() {
        std::vector<std::vector<std::set<int>>> candidates(GRID_SIZE, std::vector<std::set<int>>(GRID_SIZE, std::set<int>{1, 2, 3, 4, 5, 6, 7, 8, 9}));
        for (int row = 0; row < GRID_SIZE; ++row) {
            for (int col = 0; col < GRID_SIZE; ++col) {
                if (grid[row][col]!= 0) {
                    candidates[row][col] = std::set<int>();
                    int value = grid[row][col];
                    for (int i = 0; i < GRID_SIZE; ++i) {
                        candidates[row][i].erase(value);
                        candidates[i][col].erase(value);
                    }
                    int box_row = (row / BOX_SIZE) * BOX_SIZE;
                    int box_col = (col / BOX_SIZE) * BOX_SIZE;
                    for (int i = box_row; i < box_row + BOX_SIZE; ++i) {
                        for (int j = box_col; j < box_col + BOX_SIZE; ++j) {
                            candidates[i][j].erase(value);
                        }
                    }
                }
            }
        }
        return candidates;
    }

    bool solve() {
        std::vector<std::vector<std::set<int>>> candidates = getInference();
        for (int row = 0; row < GRID_SIZE; ++row) {
            for (int col = 0; col < GRID_SIZE; ++col) {
                if (grid[row][col] == 0) {
                    std::set<int> possible_values = candidates[row][col];
                    for (int value : possible_values) {
                        grid[row][col] = value;
                        if (solve()) {
                            return true;
                        } else {
                            grid[row][col] = 0;
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }
};

int main() {
    std::string input_str = "017903600000080000900000507072010430000402070064370250701000065000030000005601720";
    Sudoku sudoku(input_str);
    if (sudoku.solve()) {
        for (const auto& row : sudoku.grid) {
            for (int value : row) {
                std::cout << value << " ";
            }
            std::cout << std::endl;
        }
    } else {
        std::cout << "No solution found." << std::endl;
    }
    return 0;
}

File: hw1/dynamic/Sudoku.py
class Grid:
    def __init__(self):
        self.BOX_SIZE = 3
        self.GRID_SIZE = 9
        self.grid = [[0] * self.GRID_SIZE for _ in range(self.GRID_SIZE)]

    def getRow(self, row):
        return self.grid[row]

    def getColumn(self, col):
        return [self.grid[i][col] for i in range(self.GRID_SIZE)]

    def getBox(self, row, col):
        box_row_start = (row // self.BOX_SIZE) * self.BOX_SIZE
        box_col_start = (col // self.BOX_SIZE) * self.BOX_SIZE
        box = []
        for i in range(box_row_start, box_row_start + self.BOX_SIZE):
            for j in range(box_col_start, box_col_start + self.BOX_SIZE):
                box.append(self.grid[i][j])
        return box

class Sudoku(Grid):
    def __init__(self, input_str):
        super(Sudoku, self).__init__()
        self.parse(input_str)

    def parse(self, input_str):
        for row in range(self.GRID_SIZE):
            for col in range(self.GRID_SIZE):
                index = row * self.GRID_SIZE + col
                value = int(input_str[index])
                self.grid[row][col] = value

    def getInference(self):
        candidates = [[set(range(1, 10)) for _ in range(self.GRID_SIZE)] for _ in range(self.GRID_SIZE)]
        for row in range(self.GRID_SIZE):
            for col in range(self.GRID_SIZE):
                if self.grid[row][col]!= 0:
                    candidates[row][col] = set()
                    value = self.grid[row][col]
                    for i in range(self.GRID_SIZE):
                        candidates[row][i].discard(value)
                        candidates[i][col].discard(value)
                    box_row = (row // self.BOX_SIZE) * self.BOX_SIZE
                    box_col = (col // self.BOX_SIZE) * self.BOX_SIZE
                    for i in range(box_row, box_row + self.BOX_SIZE):
                        for j in range(box_col, box_col + self.BOX_SIZE):
                            candidates[i][j].discard(value)
        return candidates

    def solve(self):
        candidates = self.getInference()
        for row in range(self.GRID_SIZE):
            for col in range(self.GRID_SIZE):
                if self.grid[row][col] == 0:
                    possible_values = candidates[row][col]
                    for value in possible_values:
                        self.grid[row][col] = value
                        if self.solve():
                            return True
                        else:
                            self.grid[row][col] = 0
                    return False
        return True



if __name__ == '__main__':
    input_str = "017903600000080000900000507072010430000402070064370250701000065000030000005601720"
    sudoku = Sudoku(input_str)
    if sudoku.solve():
        for row in sudoku.grid:
            print(row)
    else:
        print("No solution found.")